torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -588,16 +646,21 @@ void TensorPipeAgent::respond(std::shared_ptr<tensorpipe::Pipe>& pipe) {
           if (futureResponseMessage->completed()) {
             decreaseCallCount(serverActiveCalls_);
             sendCompletedResponseMessage(
-                pipe, futureResponseMessage, messageId);
+                pipe, futureResponseMessage, messageId, std::move(ctx));
           } else {
             // Not complete yet
             increaseCallCount(serverActiveAsyncCalls_);
             futureResponseMessage->addCallback(
-                [this, pipe, futureResponseMessage, messageId]() mutable {
+                [this,
+                 pipe,
+                 futureResponseMessage,
+                 messageId,
+                 ctx{std::move(ctx)}]() mutable {
                   decreaseCallCount(serverActiveCalls_);
                   decreaseCallCount(serverActiveAsyncCalls_);
+                  FullDeviceContextGuard guard(ctx);
```

Comment:
I don't think a guard is needed here. From now on the streams are manipulated "explicitly" (i.e., by passing by pointers around, rather than by accessing the current ones). Was there a specific reason this was added?

Reply:
addressed